### PR TITLE
Nonnative fixes and improvements -> Dev

### PR DESCRIPTION
**Pull request by 95DDB**
Must restore marlin and poly commit dependency once merged

### CODE REVIEW

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs [hunk 1]
```diff
@@ -33,27 +35,197 @@ use std::{borrow::Borrow, vec, vec::Vec};
...
+
+
+/// Converts an unsigned big integer `bigint` into an element from the constraint field F_p by
+/// computing (bigint mod p).
+pub fn bigint_to_constraint_field<ConstraintF: PrimeField>(bigint: &BigUint) -> ConstraintF {
+    let mut val = ConstraintF::zero();
+    let mut cur = ConstraintF::one();
+    let bytes = bigint.to_bytes_be();
+
+    let basefield_256 = ConstraintF::from_repr(<ConstraintF as PrimeField>::BigInt::from(256));
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs]: Why can't we simply employ `ConstraintF::from(256)` here?

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs [hunk 2]
```diff
@@ -179,211 +569,445 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
...
+        // where 
+        // ``
+        //      surfeit(product) = log(num_limbs * (num_add(L)+1) * (num_add(R) + 1)).
+        // ``
+        // To allow for a subsequent reduction we need to assure the stricter condition 
+        // that 
+        // ``
+        //     2 * bits_per_limb + surfeit' <= CAPACITY - 2,
+        // ``
+        // with `surfeit' = log(num_limbs * (num_add(L) + 1) * (num_add(R) + 1) + num_limbs)`.
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs]: A facotr of 2 is missing, will be added with PR #164

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs [hunk 3]
```diff
@@ -179,211 +569,445 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
...
+    /// For advanced use, multiply and output the intermediate representations (without reduction)
+    /// This intermediate representations can be added with each other, and they can later be
+    /// reduced back to the `NonNativeFieldGadget`.
+    /// Assumes that
+    /// ``
+    ///     2 * bits_per_limb + surfeit' <= CAPACITY - 2,
+    /// ``
+    /// where
+    /// ``
+    ///      surfeit' = log(num_limbs^2 * (num_add(L)+1) + 1).
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs]: Shouldn't this be num_limbs + 2 * num_limbs * (num_add(L)+1), which is equivalent to the formula for `mul_without_prereduce` with num_add(R) = 0?

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs [hunk 4]
```diff
@@ -179,211 +569,445 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
...
+        }
+
+        let mut prod_limbs = Vec::new();
+        let other_limbs = Self::get_limbs_representations(other)?;
+
+        // The naive gathering of the limb representation for the product:
+        // ``
+        //      prod_limb[k] = Sum_{i+j=k} L[i] * R[j]. 
+        // ``
+        // Consumes `num_limbs^2` constraints, and `2 * num_limbs^2`
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs]: Should cost 0 constraints, since we employ `mul_by_constant` rather than `mul`

File: r1cs/gadgets/std/src/fields/nonnative/reduce.rs [hunk 5]
```diff
@@ -98,110 +76,297 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField> Reducer<SimulationF, Cons
...
+        // we need to assure the stricter condition
+        // ``
+        //     bits_per_limb + log(num_add(sum) + 3) <= CAPACITY - 3.
+        // `` 
+        Self::reduce_until_cond_is_satisfied(
+            cs,
+            elem,
+            elem_other,
+            |elem, elem_other| {
+                let sum_add = &elem.num_of_additions_over_normal_form + &elem_other.num_of_additions_over_normal_form;
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/reduce.rs]: Isn't a +1 missing in `sum_add`?

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs [hunk 6]
```diff
@@ -650,22 +1462,29 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField> FromBitsGadget<Constraint
 impl<SimulationF: PrimeField, ConstraintF: PrimeField> ToBytesGadget<ConstraintF>
     for NonNativeFieldGadget<SimulationF, ConstraintF>
 {
+    // Returns the big endian bit representation of `self mod p`.
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs]: Fix the comment: `to_bytes` should return the little-endian byte representation of `self mod p`

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs [hunk 7]
```diff
@@ -42,12 +61,12 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
         limbs.resize(2 * params.num_limbs - 1, FpGadget::<ConstraintF>::zero(cs)?);
         limbs.reverse();
 
-        let prod_of_num_of_additions =
-            other.num_of_additions_over_normal_form + &ConstraintF::one();
+        let num_add_over_normal_form =
+            BigUint::one() + &other.num_of_additions_over_normal_form;
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs]: num_add should be divided by 2^bits_per_limb, as limbs in `NonNativeFieldMulResultGadget` are 2 times bigger than in `NonNativeFieldGadget`. Btw, this fix will be introduced with PR #164.

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs [hunk 8]
```diff
@@ -56,6 +75,42 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
...
+        let normal_form_bound_ms = Pow::pow(
+            BigUint::from(2u64),
+            &(2 * (SimulationF::size_in_bits() - (params.num_limbs - 1) * params.bits_per_limb))
+        );
+        let num_add_plus_one = BigUint::one() + &self.num_add_over_normal_form;
+        let limb_bound = normal_form_bound * &num_add_plus_one;
+        let limb_bound_ms = normal_form_bound_ms * &num_add_plus_one;
+        
+        let valid_num_adds = 2 * params.bits_per_limb + ceil_log_2!(num_add_plus_one)
+             < ConstraintF::size_in_bits() - 1;
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs]: Can't this be <= instead of <?

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs [hunk 9]
```diff
@@ -33,27 +35,197 @@ use std::{borrow::Borrow, vec, vec::Vec};
...
+        let normal_form_bound = BigUint::from(2usize).pow(params.bits_per_limb as u32);
+        let normal_form_bound_ms = BigUint::from(2usize).pow(
+            (SimulationF::size_in_bits() - (params.num_limbs - 1) * params.bits_per_limb) as u32
+        );
+        let num_add_plus_one = self.num_of_additions_over_normal_form.clone() + BigUint::one();
+        let limb_bound = normal_form_bound * &num_add_plus_one;
+        let limb_bound_ms = normal_form_bound_ms * &num_add_plus_one;
+
+        let valid_num_adds = params.bits_per_limb + ceil_log_2!(num_add_plus_one)
+             < ConstraintF::size_in_bits() - 1;
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs]: Can't this be <= rather than <?

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs [hunk 10]
```diff
@@ -165,21 +281,30 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
...
             limbs: k_limbs,
-            num_of_additions_over_normal_form: self.prod_of_num_of_additions,
-            is_in_the_normal_form: false,
+            num_of_additions_over_normal_form: BigUint::from(2u32) * self.num_add_over_normal_form.clone()
+                + BigUint::one(),
             simulation_phantom: PhantomData,
         };
 
+        // alloc r in normal form. 
+        // Costs `C = len(p)` constraints.
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs]: Shoudl cost `len(p) + num_limbs` constraints due to unpacking constraints. Will be fixed in PR #164

File: r1cs/gadgets/std/src/fields/nonnative/reduce.rs [hunk 11]
```diff
@@ -98,110 +76,297 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField> Reducer<SimulationF, Cons
...
+        // ``
+        // where
+        // ``
+        //      S - 1 = Floor[
+        //          (ConstraintF::CAPACITY - 2 - (bits_per_limb + surfeit)) / shift_per_limb
+        //      ].
+        // ``
+
+        if shift_per_limb < 2 {
+            return Err(SynthesisError::Other(format!("shift_per_limb must be smaller than 2. Found: {}", shift_per_limb)));
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/reduce.rs]: Error messahe should be: "shift_per_limb must not be smaller than 2..."

File: r1cs/gadgets/std/src/fields/nonnative/reduce.rs [hunk 12]
```diff
@@ -98,110 +76,297 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField> Reducer<SimulationF, Cons
...
+    // form instead of `surfeit`.
     pub fn group_and_check_equality<CS: ConstraintSystemAbstract<ConstraintF>>(
         mut cs: CS,
+        // The additional number of bits beyond `bits_per_limb`. Hence the current
+        // limbsize is `bits_per_limb + surfeit`.
         surfeit: usize,
+        // The number of bits  
         bits_per_limb: usize,
+        // defines arity of the limb representation, i.e. `A= 2^shift_per_limb`.
+        // MUST be >= 2.
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/reduce.rs]: If there is this assumption of `shift_per_limb`, then since `shift_per_limb` is usually equal to `params.bits_per_limb`, then for safety we should avoid that `find_parameters` may choose `params.bits_per_limb = 1`, which is unlikely but not prevented with the current implementation. If this is true, then I will modify also the fixed version of `find_parameters` in PR #164 to start from `bits_per_limb = 2` rather than `bits_per_limb = 1`.

File: r1cs/gadgets/std/src/fields/nonnative/params.rs [hunk 13]
```diff
@@ -1,99 +1,99 @@
...
-pub const fn find_parameters(
+/// Finding parameters which minimize the number of constraints of a `single mul_without_prereduce()`
+/// and a subsequent `reduce()` on operand which have `surfeit = 10`, assuming that for both operands
+/// ``
+///     num_adds + 1 = 2^10 - 1.
+/// `` 
+// TODO: Let us reconsider the treatment of parameters, by introducing separate 
+// NonNativeParameters, and modifying the optimum finder to use the constraint counts
+// from the synthesizer. 
+pub(crate) const fn find_parameters(
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/params.rs]: The correct formulas to be employed in this function are found in PR #164 . Shall we move them also here or can we wait for #164 to be merged?

File: r1cs/gadgets/std/src/fields/nonnative/tests.rs [hunk 14]
```diff
@@ -1,365 +1,1120 @@
...
+            cs.ns(|| "alloc random" ),
+            rng, 
+            surfeit).unwrap();
 
-    let b_native = SimulationF::rand(rng);
-    let b = NonNativeFieldGadget::<SimulationF, ConstraintF>::alloc(cs.ns(|| "alloc b"), || {
-        Ok(b_native)
-    })
-    .unwrap();
+        println!("actual surfeit: {} ", ceil_log_2!(&a.num_of_additions_over_normal_form + BigUint::one()));
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/tests.rs]: Maybe we can compare actual surfeit with the surfeit passed to the function rather than just printing them?

File: r1cs/gadgets/std/src/fields/nonnative/tests.rs [hunk 15]
```diff
@@ -1,365 +1,1120 @@
...
+            "allocated normal form fails on check()"
+        );
+
+        a
+            .conditional_enforce_equal_without_prereduce(cs.ns(|| "non-normal form == normal form"), 
+                &a_normal_form, 
+                &Boolean::constant(true)
+            ).unwrap();
+        a_normal_form
+            .conditional_enforce_equal(cs.ns(|| "normal form == non-normal form"),
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/tests.rs]: Why don't we call `conditional_enforce_equal_without_prereduce` also here and for the subsequent call?

File: r1cs/gadgets/std/src/fields/nonnative/tests.rs [hunk 16]
```diff
@@ -1,365 +1,1120 @@
...
+                (ConstraintF::Params::CAPACITY as usize - 3 - params.bits_per_limb) as u32
+            ) 
+            - BigUint::from(2u32).pow(surfeit_bound_a)
+            - BigUint::from(3u32) 
+        ).bits() as usize - 1;
+
+        let surfeit_a = rng.gen_range(0..surfeit_bound_a);
+        // every 10-th run we create an edge cases
+        let surfeit_b = if i%10 == 0 {
+            surfeit_bound_b
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/tests.rs]: I am not sure that this is always an edge-case, as `surfeit_a` may be smaller than `surfeit_bound_a`, which is the surfeit employed to compute `surfeit_bound_b`. Shouldn't we employ the same logic of `elementary_test_add_without_prereduce` employing 2^{CAPACITY-3-bits_per_limb}-3 as an upper bound for 2^surfeit_a + 2^surfeit_b?

File: r1cs/gadgets/std/src/fields/nonnative/tests.rs [hunk 17]
```diff
@@ -1,365 +1,1120 @@
...
+        // Edge cases of mul_without_prereduce() are characterized by
+        // ``
+        //    num_limbs^2 * (num_add + 1)  + 1  
+        //                          <= 2^{CAPACITY - 2 - 2*bits_per_limb},
+        // ``
+        // or
+        // ``
+        //    num_limbs^2 * 2^surfeit_a  + 1  
+        //                          <= 2^{CAPACITY - 2 - 2*bits_per_limb},
+        // ``
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/tests.rs]: These bounds have to be updated accordingly if the bounds on `mul_by_constant_without_prereduce` are updated as suggested in the other comment.

File: r1cs/gadgets/std/src/fields/nonnative/tests.rs [hunk 18]
```diff
@@ -368,22 +1123,27 @@ fn randomized_arithmetic_test<SimulationF: PrimeField, ConstraintF: PrimeField,
...
-        num.add_in_place(cs.ns(|| format!("num += next {}", i)), &next)
-            .unwrap();
+        num_native -= &next_native;
+        let neg_next = next.negate(
+            cs.ns(||format!("nex num negate {}",i))
+        ).unwrap();
+
+        num = num.add(
+            cs.ns(|| format!("num -= next {}", i)), &neg_next
+            ).unwrap();
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/tests.rs:1199]: why using negate+add rather than sub if it is a subtraction test and there is a subsequent negation stress test?

File: r1cs/gadgets/std/src/fields/nonnative/tests.rs [hunk 19]
```diff
@@ -394,7 +1154,38 @@ fn addition_stress_test<SimulationF: PrimeField, ConstraintF: PrimeField, R: Rng
...
 }
 
-/// Tests correctness of `TEST_COUNT` many `mul_in_place` on a random instance.
+fn negation_stress_test<SimulationF: PrimeField, ConstraintF: PrimeField, R: RngCore>(rng: &mut R) {
+    let mut cs = ConstraintSystem::<ConstraintF>::new(SynthesisMode::Debug);
+
+    let num_native = SimulationF::rand(rng);
+    let mut num = NonNativeFieldGadget::alloc(cs.ns(|| "initial num"), || Ok(num_native)).unwrap();
+    for i in 0..STRESS_TEST_COUNT {
+        num = num.negate(
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/tests.rs]: Wouldn't be better to sample a new field element at each iteration rather than always negating the same two field elements?

File: r1cs/gadgets/std/src/fields/nonnative/tests.rs [hunk 20]
```diff
@@ -546,8 +1430,8 @@ fn double_stress_test_1<SimulationF: PrimeField, ConstraintF: PrimeField, R: Rng
     assert!(cs.is_satisfied());
 }
 
-/// Tests correctness of `TEST_COUNT` many steps of the recursion
-/// `x <- (x+x)*(x+x)`, starting with a random non-native `x`.
+/// Tests correctness of `STRESS_TEST_COUNT` many steps of the recursion
+/// `x <- x+x`, starting with a random non-native `x`, 
 fn double_stress_test_2<SimulationF: PrimeField, ConstraintF: PrimeField, R: RngCore>(rng: &mut R) {
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/tests.rs:1490]: Comment seems to be correct earlier, why was it modified?

File: r1cs/gadgets/std/src/fields/nonnative/tests.rs [hunk 21]
```diff
@@ -576,7 +1465,7 @@ fn double_stress_test_2<SimulationF: PrimeField, ConstraintF: PrimeField, R: Rng
     assert!(cs.is_satisfied());
 }
 
-/// Tests correctness of `TEST_COUNT` many steps of the recursion
+/// Tests correctness of `STRESS_TEST_COUNT` many steps of the recursion
 /// `x <- (x+x)*(x+x)`, starting with a random non-native `x`.  
 fn double_stress_test_3<SimulationF: PrimeField, ConstraintF: PrimeField, R: RngCore>(rng: &mut R) {
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/tests.rs:1524]: This comment should correspond to `double_stress_test_2`, here it is done x <- x+x and (x+x)^2 is also computed in each iteration to be employed in `enforce_equal`

File: r1cs/gadgets/std/src/groups/nonnative/short_weierstrass/short_weierstrass_jacobian.rs [hunk 22]
```diff
@@ -84,8 +84,7 @@ where
         &mut self,
         mut cs: CS,
     ) -> Result<(), SynthesisError> {
-        let x_squared = self.x.mul_without_reduce(cs.ns(|| "x^2"), &self.x)?;
-        //TODO: Once mul_by_constant is implemented properly we can avoid all these adds
+        let x_squared = self.x.mul_without_prereduce(cs.ns(|| "x^2"), &self.x)?;
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/groups/nonnative/short_weierstrass/short_weierstrass_jacobian.rs:87]: I would keep the ToDo in the code, it does seem to be addressed here

File: r1cs/gadgets/std/src/fields/nonnative/reduce.rs [hunk 23]
```diff
@@ -252,25 +424,95 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField> Reducer<SimulationF, Cons
...
+        // The group totals have arity `A^S = 2^{S*shift_per_limb}`.
         for (group_id, (left_total_limb, right_total_limb, num_limb_in_this_group)) in
             groupped_limb_pairs.iter().enumerate()
         {
+            // Carry and remainder are subject to the following constraints: 
+            // The quotient-remainder constraint
+            // ``
+            //    carry_in + group_total_left + shift_constant - group_total_right 
+            //          == carry * A^S + 0 + (shift_constant % A^S),
+            // ``
```
Comment (nicholas-mainardi) [inline r1cs/gadgets/std/src/fields/nonnative/reduce.rs:537]: The constraint should be: `carry_in + group_total_left + shift_constant - group_total_right == carry* A^S + 0 +(accumulated_extra + shift_constant % A^S)`. Maybe add also a brief comment to intuitively explain what the variable `accumulated_extra` represents (e.g., the carries of the shift constants)

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs (hunk 1 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs]: I agree, would be more straight-forward.

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs (hunk 9 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs]: I agree. This function is supposed to  assures that the limb size is at most CAPACITY.

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs (hunk 2 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs]: You are right. The correct formula is
 ```
    surfeit' = log(num_limbs + num_adds(prod) + 1) =
                = log(num_limbs + 2 * num_limbs * (num_add(L) + 1) * (num_add(R) + 1))
                = log(num_limbs * (1 + 2 * (num_add(L) + 1) * (num_add(R) + 1))) 
```

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs (hunk 3 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs]: Thanks for pointing this out. You are right.

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs (hunk 4 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs]: Right. A left-over from copy-paste.

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs (hunk 6 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs]: Agree!

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs (hunk 7 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs]: Agree!

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs (hunk 8 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs]: As above, you are right.

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs (hunk 10 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs]: Agree. This is exactly the gap in the constraint counts of the parameter finder.

File: r1cs/gadgets/std/src/fields/nonnative/reduce.rs (hunk 5 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/reduce.rs]: It seems to me that the `add_without_prereduce()` has a too strict reducible bound
```
     bits_per_limb + log(num_add(L) + num_add(R) + 4) <= CAPACITY - 3.
```
It should be as commented in Line 128.

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs [hunk 24]
```diff
@@ -88,54 +260,270 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
...
         let params = get_params(SimulationF::size_in_bits(), ConstraintF::size_in_bits());
+        let surfeit = ceil_log_2!(BigUint::from(4usize) + &self.num_of_additions_over_normal_form + &other.num_of_additions_over_normal_form);
+        
+        if params.bits_per_limb + surfeit > ConstraintF::Params::CAPACITY as usize - 3 {
+            return Err(SynthesisError::Other(
+                format!("Security bound exceeded for add_without_prereduce. Max: {}, Actual: {}", 
+                ConstraintF::Params::CAPACITY as usize - 3, 
+                params.bits_per_limb + surfeit))
+            );
+        }
```
Comment (UlrichHaboeck75) [inline r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs]: The reducible condition should be 
```
     bits_per_limb + log(num_add(L) + num_add(R) + 3) <= CAPACITY - 3.
```

File: r1cs/gadgets/std/src/fields/nonnative/reduce.rs (hunk 12 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/reduce.rs]: Agree, that would be a good strategy. It is not useful if `find_parameters` outputs parameters which do not work.

File: r1cs/gadgets/std/src/fields/nonnative/reduce.rs (hunk 23 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/reduce.rs:537]: I agree. This was not well explained in the comments. I will add the explanation.

File: r1cs/gadgets/std/src/fields/nonnative/mod.rs [hunk 25]
```diff
@@ -10,14 +9,17 @@
 //!     result of multiplication, which is hidden from the `FieldGadget` interface
 //!     and is left for advanced users who want better performance.
 //! DISCLAIMER: THIS LIBRARY IS EXPERIMENTAL AND NEEDS TO UNDERGO A MORE IN-DEPTH REVIEW
```
Comment (UlrichHaboeck75) [inline r1cs/gadgets/std/src/fields/nonnative/mod.rs]: Let us remove this disclaimer.

File: r1cs/gadgets/std/src/fields/nonnative/reduce.rs (hunk 23 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/reduce.rs:537]: Done in in PR #164.

File: r1cs/gadgets/std/src/fields/nonnative/tests.rs (hunk 14 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/tests.rs]: Agree.

File: r1cs/gadgets/std/src/fields/nonnative/tests.rs (hunk 15 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/tests.rs]: You are right. All these calls should be without prereduction.

File: r1cs/gadgets/std/src/fields/nonnative/tests.rs (hunk 16 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/tests.rs]: Let us do it!

File: r1cs/gadgets/std/src/fields/nonnative/tests.rs (hunk 17 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/tests.rs]: Agree!

File: r1cs/gadgets/std/src/fields/nonnative/tests.rs (hunk 18 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/tests.rs:1199]: Good point. Let us change it.

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs (hunk 24 above)
Comment (nicholas-mainardi) [reply r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs]: I think it is right, if you want a reducible sum, then you have to satisfy `bits_per_limb + log(num_add(sum) + 3) <= CAPACITY - 3`, but `num_add(sum) = num_add(L)+num_add(R)+1`, hence overall you get the condition `bits_per_limb + log(num_add(L) + num_add(R) + 4) <= CAPACITY - 3`. So I think that the bound here is correct while in prereduce function `num_add(sum)` is wrongly computed as `num_add(L) + num_add(R)` rather than `num_add(L)+num_add(R)+1`

File: r1cs/gadgets/std/src/fields/nonnative/tests.rs (hunk 19 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/tests.rs]: Stress tests are supposed to test larger arithmetic "networks", in this case a large computation using neg. The neg does interally a substraction, a double neg does not produce the same limb presentation we started off.

File: r1cs/gadgets/std/src/fields/nonnative/tests.rs (hunk 21 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/tests.rs:1524]: Let us update the comment!

**Review by UlrichHaboeck75 — CHANGES_REQUESTED** [review]
Beside the comments I addressed, ok.

File: r1cs/gadgets/std/src/fields/nonnative/reduce.rs (hunk 12 above)
Comment (nicholas-mainardi) [reply r1cs/gadgets/std/src/fields/nonnative/reduce.rs]: Ok then I will add the fix directly to PR #164, since the correct implementations of `find_parameters` are in that PR

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs (hunk 4 above)
Comment (nicholas-mainardi) [reply r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs]: Ok, will be fixed in PR #164, where the `mul_by_constant_without_prereduce` is refactored with the `mul_without_prereduce` functions, so no need to modify it here

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs (hunk 3 above)
Comment (nicholas-mainardi) [reply r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs]: Ok, will be fixed in PR https://github.com/HorizenOfficial/ginger-lib/pull/164, where the `mul_by_constant_without_prereduce` is refactored with the `mul_without_prereduce` functions, so no need to modify it here

File: r1cs/gadgets/std/src/fields/nonnative/mod.rs (hunk 25 above)
Comment (nicholas-mainardi) [reply r1cs/gadgets/std/src/fields/nonnative/mod.rs]: Already done in PR #164

File: r1cs/gadgets/std/src/fields/nonnative/tests.rs (hunk 17 above)
Comment (nicholas-mainardi) [reply r1cs/gadgets/std/src/fields/nonnative/tests.rs]: The part of the comment related to `mul_without_prereduce` will be removed in PR #164, as those bounds are not employed in this test (`mul_by_constant` is employed rather than `mul_without_prereduce`)

File: r1cs/gadgets/std/src/fields/nonnative/reduce.rs (hunk 5 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/reduce.rs]: Sorry, I was mistaken: the reducability bound of `add_without_prereduce()` is fine. I simply missed the fact that 
```
   num_add(sum) = num_add(L) + num_add(R) + 1.
``` 
So you are right, there is a `+1` missing.

Comment (nicholas-mainardi) [reply r1cs/gadgets/std/src/fields/nonnative/reduce.rs]: Done in PR #164 .

File: r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs (hunk 24 above)
Comment (UlrichHaboeck75) [reply r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs]: Understood. You are right.
